File: hotel-management-backend/src/main/java/com/example/hotel_management/service/ReservationService.java
package com.example.hotel_management.service;

import com.example.hotel_management.dto.ReservationDTO;
import com.example.hotel_management.dto.HotelStatsDTO;
import com.example.hotel_management.entity.Reservation;
import com.example.hotel_management.enums.ReservationStatus;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

/**
 * Service interface for managing hotel reservations.
 * Provides methods for creating, updating, retrieving, and managing reservation status.
 */
public interface ReservationService {
    
    /**
     * Creates a new reservation.
     * Validates dates, calculates total price, and checks room availability.
     *
     * @param reservation The reservation to create
     * @return The created reservation with generated ID and calculated price
     * @throws IllegalArgumentException if reservation dates are invalid
     * @throws IllegalStateException if the room is not available
     */
    ReservationDTO createReservation(Reservation reservation);

    /**
     * Updates an existing reservation.
     * Recalculates price if dates change and validates new data.
     *
     * @param id The ID of the reservation to update
     * @param reservation The updated reservation data
     * @return The updated reservation
     * @throws EntityNotFoundException if reservation not found
     * @throws IllegalArgumentException if updated dates are invalid
     * @throws IllegalStateException if the room is not available for new dates
     */
    ReservationDTO updateReservation(Long id, Reservation reservation);

    /**
     * Retrieves a reservation by its ID.
     *
     * @param id The ID of the reservation
     * @return The reservation
     * @throws EntityNotFoundException if reservation not found
     */
    ReservationDTO getReservationById(Long id);

    /**
     * Retrieves all reservations in the system.
     *
     * @return List of all reservations
     */
    List<ReservationDTO> getAllReservations();

    /**
     * Retrieves all reservations for a specific guest.
     *
     * @param guestId The ID of the guest
     * @return List of reservations for the guest
     */
    List<ReservationDTO> getReservationsByUserId(Long guestId);

    /**
     * Retrieves all reservations for a specific hotel.
     *
     * @param hotelId The ID of the hotel
     * @return List of reservations for the hotel
     */
    List<ReservationDTO> getReservationsByHotelId(Long hotelId);

    /**
     * Retrieves upcoming reservations for a guest.
     * Only returns reservations with check-in time after current time.
     *
     * @param guestId The ID of the guest
     * @return List of upcoming reservations
     */
    List<ReservationDTO> getUpcomingReservations(Long guestId);

    /**
     * Retrieves current active reservations for a guest.
     * Returns reservations where current time is between check-in and check-out.
     *
     * @param guestId The ID of the guest
     * @return List of current reservations
     */
    List<ReservationDTO> getCurrentReservations(Long guestId);

    /**
     * Updates the status of a reservation based on its check-in and check-out times.
     * Changes status to CHECKED_IN if check-in time has passed and status is CONFIRMED.
     * Changes status to CHECKED_OUT if check-out time has passed and status is CHECKED_IN.
     *
     * @param id The ID of the reservation to update
     */
    void updateReservationStatusBasedOnTime(Long id);

    /**
     * Updates the status of all reservations based on their check-in and check-out times.
     * This method is scheduled to run periodically.
     */
    void updateReservationStatuses();

    /**
     * Cancels a reservation.
     * Sets the reservation status to CANCELLED.
     *
     * @param id The ID of the reservation to cancel
     * @throws EntityNotFoundException if reservation not found
     * @throws IllegalStateException if reservation is already cancelled
     */
    void cancelReservation(Long id);

    boolean isRoomAvailable(Long roomId, LocalDateTime checkIn, LocalDateTime checkOut);

    @Transactional
    ReservationDTO updateReservationStatus(Long id, ReservationStatus status);

    /**
     * Get comprehensive statistics for a specific hotel
     *
     * @param hotelId The ID of the hotel
     * @return HotelStatsDTO containing various statistics
     */
    HotelStatsDTO getHotelStats(Long hotelId);
}

File: hotel-management-frontend/src/services/reservation.service.ts
import { api } from './api.config';
import { Reservation, CreateReservationDTO, ReservationStatus } from '../types/reservation.type';

export class ReservationService {
    private static readonly BASE_PATH = '/reservations';

    static async createReservation(reservation: CreateReservationDTO): Promise<Reservation> {
        const response = await api.post(this.BASE_PATH, reservation);
        return response.data;
    }

    static async updateReservation(id: number, reservation: Reservation): Promise<Reservation> {
        const response = await api.put(`${this.BASE_PATH}/${id}`, reservation);
        return response.data;
    }

    static async updateReservationDTO(id: number, reservation: CreateReservationDTO): Promise<Reservation> {
        const response = await api.put(`${this.BASE_PATH}/${id}`, reservation);
        return response.data;
    }

    static async getReservation(id: number): Promise<Reservation> {
        const response = await api.get(`${this.BASE_PATH}/${id}`);
        return response.data;
    }

    static async getAllReservations(): Promise<Reservation[]> {
        const response = await api.get(this.BASE_PATH);
        return response.data;
    }

    static async getReservationsByGuest(guestId: number): Promise<Reservation[]> {
        const response = await api.get(`${this.BASE_PATH}/guest/${guestId}`);
        return response.data;
    }

    static async getReservationsByHotel(hotelId: number): Promise<Reservation[]> {
        const response = await api.get(`${this.BASE_PATH}/hotel/${hotelId}`);
        return response.data;
    }

    static async getUpcomingReservations(guestId: number): Promise<Reservation[]> {
        const response = await api.get(`${this.BASE_PATH}/guest/${guestId}/upcoming`);
        return response.data;
    }

    static async getCurrentReservations(guestId: number): Promise<Reservation[]> {
        const response = await api.get(`${this.BASE_PATH}/guest/${guestId}/current`);
        return response.data;
    }

    static async cancelReservation(id: number): Promise<void> {
        await api.post(`${this.BASE_PATH}/${id}/cancel`);
    }

    static async updateReservationStatus(id: number, status: ReservationStatus): Promise<Reservation> {
        try {
            console.log('Updating reservation status:', { id, status });
            const response = await api.patch(`${this.BASE_PATH}/${id}/status`, { status });
            const updatedReservation = response.data;
            console.log('Updated reservation:', updatedReservation);

            // Update the room status based on the new reservation status
            if (updatedReservation.room && updatedReservation.room.id) {
                console.log('Updating room status for room:', updatedReservation.room.id);
                await this.updateRoomStatusForReservation(updatedReservation.room.id, status);
            } else {
                console.warn('No room ID found in reservation:', updatedReservation);
            }

            return updatedReservation;
        } catch (error) {
            console.error('Error in updateReservationStatus:', error);
            throw error;
        }
    }

    private static async updateRoomStatusForReservation(roomId: number, reservationStatus: ReservationStatus): Promise<void> {
        try {
            console.log('Making room status update request:', {
                roomId,
                reservationStatus
            });
            // Remove /api prefix since it's already in the baseURL
            const response = await api.patch(`/rooms/${roomId}/status/by-reservation`, { 
                reservationStatus 
            });
            console.log('Room status update response:', response.data);
        } catch (error) {
            console.error('Error updating room status:', error);
            throw error;
        }
    }

    static async getHotelReservationStats(hotelId: number): Promise<{
        totalBookings: number;
        activeGuests: number;
        availableRooms: number;
        monthlyRevenue: number;
        recentBookings: Reservation[];
        roomTypeStats: { type: string; percentage: number }[];
        upcomingCheckouts: Reservation[];
    }> {
        try {
            console.log('Fetching stats for hotel:', hotelId);
            const response = await api.get(`${this.BASE_PATH}/hotel/${hotelId}/stats`);
            console.log('Hotel stats response:', response.data);
            return response.data;
        } catch (error) {
            console.error('Error fetching hotel stats:', error);
            throw error;
        }
    }
}
